pkg/receive/multitsdb.go

```diff
@@ -647,6 +660,48 @@ func (t *MultiTSDB) TenantAppendable(tenantID string) (Appendable, error) {
 	return tenant.readyStorage(), nil
 }
 
+func (t *MultiTSDB) SetHashringConfig(cfg []HashringConfig) error {
+	t.hashringConfigs = cfg
+
+	// If a tenant's already existed in MultiTSDB, update its label set
+	// from the latest []HashringConfig.
+	// In case one tenant appears in multiple hashring configs,
+	// only the label set from the first hashring config is applied.
+	// This is the same logic as startTSDB.
+	updatedTenants := make([]string, 0)
+	for _, hc := range t.hashringConfigs {
+		for _, tenantID := range hc.Tenants {
+			if slices.Contains(updatedTenants, tenantID) {
+				continue
+			}
+			if t.tenants[tenantID] != nil {
+				updatedTenants = append(updatedTenants, tenantID)
+
+				lset := labelpb.ExtendSortedLabels(t.labels, labels.FromStrings(t.tenantLabelName, tenantID))
+
+				if hc.ExternalLabels != nil {
+					extendedLset, err := extendLabels(lset, hc.ExternalLabels, t.logger)
+					if err != nil {
+						t.mtx.Lock()
+						delete(t.tenants, tenantID)
```

Comment:
This is dangerous because a misconfiguration in the hashring config can cause a tenant to get deleted. I suggest we just return the error in this case.

Reply:
done